File: src/test/js/formulas/math/QUOTIENTTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7Exception } from "../../../../main/js/errors/F7Exception";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { QUOTIENT } from "../../../../main/js/formulas/math/QUOTIENT";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("QUOTIENT", function () {
  it("should do normal operations", function () {
    assert.deepEqual(QUOTIENT.SELF.run(null, 10, 10), 1);
    assert.deepEqual(QUOTIENT.SELF.run(null, 10, 2), 5);
    assert.deepEqual(QUOTIENT.SELF.run(null, 10, 3), 3);
    assert.deepEqual(QUOTIENT.SELF.run(null, 0, 1628736813.2), 0);
    assert.deepEqual(QUOTIENT.SELF.run(null, 218637221.22, 2876.111), 76018);
    assert.deepEqual(QUOTIENT.SELF.run(null, 9, -4), -2);
    assert.deepEqual(QUOTIENT.SELF.run(null, -9, -4), 2);
    assert.deepEqual(QUOTIENT.SELF.run(null, -9, 4), -2);
  });

  it("should do string conversion", function () {
    assert.deepEqual(QUOTIENT.SELF.run(null, "10", "10"), 1);
    assert.deepEqual(QUOTIENT.SELF.run(null, "10", "2"), 5);
    assert.deepEqual(QUOTIENT.SELF.run(null, "10", "3"), 3);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(
      (QUOTIENT.SELF.run(null, 10, new ValueException()) as F7Exception).name,
      F7ExceptionName.VALUE
    );
    assert.deepEqual(
      (QUOTIENT.SELF.run(null, new ValueException(), 10) as F7Exception).name,
      F7ExceptionName.VALUE
    );
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new QUOTIENT(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(10);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.G19_RANGE).returns(3);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE, CommonModels.G19_RANGE), 3);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.G19_RANGE));
    assert.equal(collateralLookup.callCount, 2);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 10).add(0, 1, "A").build();
    const two = Grid.builder().add(0, 0, 3).add(0, 1, "B").build();
    assert.deepEqual(QUOTIENT.SELF.run(null, one, two), 3);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((QUOTIENT.SELF.run(null, "Too few") as NAException).name, F7ExceptionName.NA);
    assert.deepEqual(
      (QUOTIENT.SELF.run(null, "A", "B", "Too many") as NAException).name,
      F7ExceptionName.NA
    );
  });
});

File: src/test/java/io/protobase/f7/formulas/math/QUOTIENTTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.DivException;
import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class QUOTIENTTest extends TestFormula {
  @Test
  public void testApply_cleanDivision() {
    assertThat(QUOTIENT.SELF.apply(null, 10.0, 10.0)).isEqualTo(1.0);
    assertThat(QUOTIENT.SELF.apply(null, 10.0, 2.0)).isEqualTo(5.0);
    assertThat(QUOTIENT.SELF.apply(null, 10.0, 3.0)).isEqualTo(3.0);
    assertThat(QUOTIENT.SELF.apply(null, 0.0, 1628736813.2)).isEqualTo(0.0);
    assertThat(QUOTIENT.SELF.apply(null, 218637221.22, 2876.111)).isEqualTo(76018.0);
    assertThat(QUOTIENT.SELF.apply(null, 9.0, -4.0)).isEqualTo(-2.0);
    assertThat(QUOTIENT.SELF.apply(null, -9.0, -4.0)).isEqualTo(2.0);
    assertThat(QUOTIENT.SELF.apply(null, -9.0, 4.0)).isEqualTo(-2.0);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(QUOTIENT.SELF.apply(null, "10.0", "10.0")).isEqualTo(1.0);
    assertThat(QUOTIENT.SELF.apply(null, "10.0", "2.0")).isEqualTo(5.0);
    assertThat(QUOTIENT.SELF.apply(null, "10.0", "3.0")).isEqualTo(3.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(QUOTIENT.SELF.apply(null,
        Grid.builder().add(0, 0, 6.0).add(0, 1, "Don't mind me.").build(),
        Grid.builder().add(0, 0, 2.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(3.0);
  }

  @Test
  public void test_lookup() {
    QUOTIENT F = new QUOTIENT(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    when(collateralLookup.apply(A1, G19_RANGE)).thenReturn(2.0);
    assertThat(F.apply(A1, M22_RANGE, G19_RANGE)).isEqualTo(5.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verify(collateralLookup).apply(A1, G19_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(QUOTIENT.SELF.apply(null, 10.0, new ValueException())).isEqualTo(new ValueException());
    assertThat(QUOTIENT.SELF.apply(null, new ValueException(), 10.0)).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(QUOTIENT.SELF.apply(null, "Too few")).isEqualTo(new NAException());
    assertThat(QUOTIENT.SELF.apply(null, "A", "B", "Too many")).isEqualTo(new NAException());
  }

  @Test
  public void testApply_errorDivideByZero() {
    assertThat(QUOTIENT.SELF.apply(null, 328467.0, 0.0)).isEqualTo(new DivException());
    assertThat(QUOTIENT.SELF.apply(null, 0.0, 0.0)).isEqualTo(new DivException());
  }
}
